[pytorch][counters] Pybind for WaitCounter

andriigrynenko opened the pull request:
Summary:
Basic pybind integration for WaitCounter providing a guard API.
Also fixes broken copy/move constructor in WaitGuard (it wasn't really used with the macro-based C++ API).

Test Plan: unit test

Differential Revision: D60557660

albanD commented on test/allowlist_for_publicAPI.json:
```diff
@@ -2101,6 +2101,7 @@
     "Event",
     "EventHandlerHandle",
     "Stat",
+    "WaitCounter",
```
Adding new entries in this file is not allowed.
Please follow the instructions from the error message to properly expose and document this class if it should be public. Or make it private properly (add a _ in front) if you don't want it to be public.

andriigrynenko replied:
Hm, I'm not sure I understand this policy. The rest of the types in this module are using the allow list, how exactly is this new type special?

andriigrynenko replied:
Also I see entries added to this file recently - https://github.com/pytorch/pytorch/commit/54b0006cb232f798281397b2261101625444c79b

andriigrynenko committed:
[pytorch][counters] Pybind for WaitCounter (#132357)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/132357

Basic pybind integration for WaitCounter providing a guard API.
Also fixes broken copy/move constructor in WaitGuard (it wasn't really used with the macro-based C++ API).

Test Plan: unit test

Reviewed By: jamesperng, asiab4

Differential Revision: D60557660
